### PR TITLE
feat(form-v2): email/storage mode respondent rollout

### LINKED ISSUE #4265: Respondent rollout based on email vs storage mode instead of auth
Auth/no-auth differentiation should be changed to email/storage mode

### PR DESCRIPTION
**Pull request by wanlingt**
## Problem
We want respondent rollout for public forms to be based on email/storage mode instead of auth/no-auth

Closes #4265 

## Solution
Replace auth/no-auth checks with storage/email mode for public forms. The % thresholds for email mode and storage public mode forms are set independently. Depending on these thresholds, a randomly chosen chosen selection of public forms will be served the React app.

**Bug Fixes**:
- If the % threshold is set to 0, delete `v2-respondent-ui` and `v2-admin-ui` cookies from localStorage(if they exist). This prevents an infinite redirection between angular and react routes, and allows us to "hard-rollback" the react rollout.

**Breaking Changes** 

- [x] No - this PR is backwards compatible  

## Tests
- [ ]  Change the storage mode rollout percentage to 100. Delete the `v2-admin-ui` cookie if it exists (as this will override `v2-respondent-ui`). Navigate to a storage-mode public form, you should be shown the React app. The `v2-respondent-ui` cookie should be set to react
- [ ] Change the storage mode rollout percentage to 0. Set either `v2-admin-ui` or `v2-respondent-ui` cookies with the value `react`. Navigate to a storage-mode public form, you should be shown the AngularJS app. The `v2-respondent-ui` and `v2-admin-ui` cookies should **no longer exist**.
- [ ] Repeat the same 2 tests for email-mode public forms

## Deploy Notes

**New environment variables**:

- `REACT_MIGRATION_RESP_ROLLOUT_EMAIL`: number between 0 and 100
- `REACT_MIGRATION_RESP_ROLLOUT_STORAGE`: number between 0 and 100

### CODE REVIEW

**Review by karrui — APPROVED** [review]
lgtm.

>Change the storage mode rollout percentage to 0. Delete the v2-respondent-ui cookie if it exists.

should we check if a rollout percentage of `0` will always ignore the cookie value and show AngularJS? this allows us to "hard-rollback" the rollout to fix any bugs we find

**Comment by wanlingt** [comment]
> lgtm.
> 
> > Change the storage mode rollout percentage to 0. Delete the v2-respondent-ui cookie if it exists.
> 
> should we check if a rollout percentage of `0` will always ignore the cookie value and show AngularJS? this allows us to "hard-rollback" the rollout to fix any bugs we find

yes thanks for raising this! I realised the infinite redirect issue still existed if a ui cookie was present + we set the rollout percentage to `0`. I've added some lines to delete the cookie if the rollout percentage is set to `0` to ensure there's a "hard rollback", and also amended the tests.
cc @timotheeg 

(Re-requesting review because of this change)

File: src/app/modules/react-migration/react-migration.controller.ts [hunk 1]
```diff
@@ -91,6 +91,11 @@ export const serveForm: ControllerHandler<
     // Check the rollout value first, if it's 0, react is DISABLED
     // And we ignore cookies entirely!
     showReact = false
+    // Delete existing cookies to prevent infinite redirection
+    if (req.cookies) {
+      res.cookie(config.reactMigration.respondentCookieName, '', { maxAge: 0 })
+      res.cookie(config.reactMigration.adminCookieName, '', { maxAge: 0 })
```
Comment (timotheeg) [inline src/app/modules/react-migration/react-migration.controller.ts]: Use the [`res.clearCookie()`](https://expressjs.com/en/api.html#res.clearCookie) method instead. Same comment for call below.

**Review by timotheeg — APPROVED** [review]
lgtm

**Pull request merged**